### PR TITLE
Batching based on estimated message size

### PR DESCRIPTION
**Pull request by SeanFeldman**
Replacing ASB native API for `BrokeredMessage.Size` with estimated message size.

Estimated size is based on the following:
- Raw body size in bytes
- Custom headers size (keys and values) in bytes
- Guestimated standard properties size (for strings assumed size is 256 bytes) in bytes
- Additional 10% to take into consideration serialization and internals added by ASB upon sending

**Some stats**

Sending a message with a specific payload, with added standard NSB headers, and _no_ exception headers led to the following results:

| Single Payload Size   |.Size   | .EstimatedSize  | Increase |
|---|---:|---:|:---:|
|0K  | 1,474  | 3,232 | 119% |
|1K   | 2,807  | 4,754 | 69% |
| 10K  | 15,147  | 18,271 | 20% |
| 100K  | 137,977  | 153,439 | 11% |
| 170K  | 233,550  | 258,568 | 11% |

Based on the numbers, small messages will not be batched very well. Perhaps, some of the assumptions need to be reviewed (estimated size of the standard properties and additional percentage added to the overall size that might be too aggressive).

@Particular/azure-maintainers please review

Connect to #86

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@Particular/azure-maintainers ping :smiley:

**Comment by SeanFeldman** [comment]
Same test with 5% gives extra 10K for payload (up to 180K)

| Single Payload Size   |.Size   | .EstimatedSize  | Increase |
|---|---:|---:|:---:|
|0K  | 1,426  | 3,088 | 117% |
|1K   | 2,859 | 4,538 | 59% |
| 10K  | 15,147 | 17,440 | 16% |
| 100K  | 138,029 | 146,464 | 7% |
| 170K  | 233,550 | 246,815 | 6% |
| 180K  | 247,201 | 261,149 | 6% |

**Comment by yvesgoeleven** [comment]
@SeanFeldman I'd prefer to let a human tune the %, make it (or the formula to compute if you decide to go for the adjusted estimate) configurable?

File: src/Transport/Sending/Batcher.cs [hunk 1]
```diff
@@ -110,5 +110,40 @@ public class BatchedOperation
...
+                                         4 + // State
+                                         8 + // TimeToLive +
+                                         assumeSize + // To +
+                                         assumeSize;  // ViaPartitionKey;
+
+            var headers = Message.Headers.Sum(property => GetStringSizeInBytes(property.Key) + GetStringSizeInBytes(property.Value));
+            var bodySize = Message.Body.Length;
+            var total = standardPropertiesSize + headers + bodySize;
+
+            const double addTenPercent = 1.1;
```
Comment (yvesgoeleven) [inline src/Transport/Sending/Batcher.cs]: Add this to config, so a user can fine tune it for his scenario. Question: if body type is stream, would this also be necessary?

**Comment by SeanFeldman** [comment]
@yvesgoeleven changed to set padding % via configuration. Default set to 5%.

**Comment by SeanFeldman** [comment]
Results of running with transport encoding set to `stream` @yvesgoeleven.
Numbers seem to be very close to `byte[]` results

| Single Payload Size   |.Size   | .EstimatedSize  | Increase |
|---|---:|---:|:---:|
|0K  | 1,401  | 3,101 | 122% |
|1K   | 2,769 | 4,538 | 64% |
| 10K  | 15,057 | 17,440 | 16% |
| 100K  | 137,937 | 146,464 | 7% |
| 170K  | 233,509 | 246,815 | 6% |
| 180K  | 247,161 | 261,149  | 6% |

**Pull request merged**